_posts/2020-09-01-java-giants.md

```diff
@@ -0,0 +1,147 @@
+![get_c4](/assets/images/2020-09-14-get-c4.png)
+Figure 3:  get() with same bytecode and new optimized assembly
+
+Now, if you stare long enough at the assembly, you'll notice several things about this rewrite:
+1.  The JIT is no longer bothering to load member variables into memory in preparation for a call. 
+2.  The JIT is no longer even calling our child method.  It copied the relevant assembly into get() itself (an optimization called "inlining", as called out in the popup). 
+3.  The JIT realized our child method is completely stupid and always returns the same answer.
+4.  The JIT has put the answer it always gets from our child method in memory address 0xffffffffffff7fae0 for get() to use.
+5.  The JIT assumes this memory address contents will match our member variable "answer".  If ever wrong, it will jump to the "uncommon trap" area of code and call the original child method.
+
+Note too that the bytecode is unchanged.   So, again, if you try to debug the new optimized code of "get()", you'll step through the (unchanged) bytecode and fail to reproduce the exact same behavior.  
+
+## Conclusion
+This post was about an infinite loop puzzle, but also much more.   As software engineers, we often search for magic incantations (code snippets) from online communities for our immediate problem at hand, enjoy introductory videos for the latest technologies or concepts, and certainly (hopefully) peruse online manuals for the same.   However, without studying the software engineering "giants", you will only learn to solve the issues that *you know about*.   You will miss learning the solutions to all the unknown issues you have yet to encounter, often interweaving concepts you have yet to come across.  For me, at the time I was first studying Hawkins, this puzzle was just one such example, and Hawkins was just one of several "giants" introducing me to the vast amount I did not know.  Enjoy studying my giants above, or discover your own, and amaze yourself with how much farther you begin to see.
+
+## Postscript
+Oh, I did not mention my quote yet – why did I include that?    Just because we can stand on the shoulders of giants does not mean we will not stumble now and then, and that is perfectly okay.   **Carol Dweck** is a giant in the field of empowering people to worry more about achievements and glorious struggles rather than always "appearing smart".   Her concept is called the “Growth Mindset” and it is legitimately awesome.   I love her 10 minute video [How to Help Every Child Fulfil Their Potential](https://youtu.be/Yl9TVbAal5s) as an (animated!) intro as well as her book “The New Psychology of Success”.    Her “Growth Mindset” is a part of Blue Yonder culture, as touted from CEO down, and that culture is one of the many reasons why 20+ years later I still love coming to work here every day.
```

Comment:
This one I didn't really get...probably because focus is depleeted already...but the video is just awesome...could that even be a separate blog post? learning culture, even blameless postmortems? But it is also fine to just keep it here

Reply:
Thank you, I made changes based on this comment.  I'm just marking the ones I know I addressed.  See the new article for details.